tfprotov5/internal/toproto/diagnostic_test.go

```diff
@@ -0,0 +1,46 @@
+package toproto
+
+import (
+	"testing"
+)
+
+func TestForceValidUTF8(t *testing.T) {
+	tests := []struct {
+		Input string
+		Want  string
+	}{
+		{
+			"hello",
+			"hello",
+		},
+		{
+			"こんにちは",
+			"こんにちは",
+		},
+		{
+			"baﬄe", // NOTE: "ﬄ" is a single-character ligature
+			"baﬄe", // ligature is preserved exactly
+		},
+		{
+			"wé́́é́́é́́!", // NOTE: These "e" have multiple combining diacritics
+			"wé́́é́́é́́!", // diacritics are preserved exactly
+		},
+		{
+			"😸😾", // Astral-plane characters
+			"😸😾", // preserved exactly
+		},
+		{
+			"\xff\xff",     // neither byte is valid UTF-8
+			"\ufffd\ufffd", // both are replaced by replacement character
+		},
```

Comment:
Nit: It may also be good to show that intermixed invalid characters are replaced as expected (which they currently are):

```suggestion
		},
		{
			"t\xffe\xffs\xfft",       // intermixed invalid UTF-8
			"t\ufffde\ufffds\ufffdt", // all are replaced by replacement character
		},
```